tensorflow-ops/tests/GradientTest.hs

```diff
@@ -164,6 +167,117 @@ testMaxGradient = testCase "testMaxGradient" $ do
         TF.gradients y [x] >>= TF.run
     V.fromList [0, 0, 1, 0, 0 :: Float] @=? dx
 
+testConcatGradient :: Test
+testConcatGradient = testCase "testConcatGradient" $ do
+    [dv,dv'] <- TF.runSession $ do
+        v  <- TF.render $ TF.vector [1 :: Float]
+        v' <- TF.render $ TF.vector [2 :: Float]
+        let y = TF.concat (TF.scalar 0) [ v, v' ]
+        TF.gradients y [v,v'] >>= TF.run
+    V.fromList [1 :: Float] @=? dv
+    V.fromList [1 :: Float] @=? dv'
+    [dw,dw'] <- TF.runSession $ do
+        w  <- TF.render $ TF.vector [1,2,3,4 :: Float]
+        w' <- TF.render $ TF.vector [5,6,7,8 :: Float]
+        let y = TF.concat (TF.scalar 0) [ w, w', w ]
+        TF.gradients y [w,w'] >>= TF.run
+    V.fromList [2,2,2,2 :: Float] @=? dw
+    V.fromList [1,1,1,1 :: Float] @=? dw'
+
+-- This test checks that the gradient of a concat op
+--   along the second dimension is as expected.
+--   This test is a port of ConcatTest._testGradientsSimple from
+--   tensorflow/tensorflow/compiler/tests/concat_ops_test.py
+testConcatGradientSimple :: Test
+testConcatGradientSimple = testCase "testConcatGradientSimple" $ do
+    let shapes     = [[10,x,2] | x <- [1,2,6]]
+    (inputGrads :: [[Float]]) <- forM shapes $ \shape ->
+       replicateM (List.product shape) randomIO
+    (inputs :: [[Float]]) <- forM shapes $ \shape ->
+       replicateM (List.product shape) randomIO
+    dinputs <- TF.runSession $ do
+        inputTensors <- forM (inputs `zip` shapes) $ \(input,shape) ->
+                          TF.render $ TF.constant (TF.Shape shape) input
+        inputGradTensors <- forM (inputGrads `zip` shapes) $ \(inputGrad, shape) ->
+                               TF.render $ TF.constant (TF.Shape shape) inputGrad
+        inputGradTensor <- TF.render $ TF.concat (TF.scalar 1) inputGradTensors
+        inputTensor <- TF.render $ TF.concat (TF.scalar 1) inputTensors
+        output <- TF.render $ inputTensor `TF.mul` inputGradTensor
+        TF.gradients output inputTensors >>= TF.run
+    (V.fromList <$> inputGrads) @=? dinputs
```

Comment:
Yes, fair enough. I've refactored out the common parts and also expanded the tests to cover negative indices.